copilot-chat-copilot.el

```diff
@@ -327,35 +327,93 @@
+  ;;    processing continues with "data: {...}\n\ndata: [D". Thus, `callback' is called with "work",
+  ;;    the next line skipped, and then "data: [D" saved to `copilot-chat-last-data'.
+  ;;
+  ;; 3. With segment 3, `copilot-chat-last-data' is prepended to `string', resulting in a value of
+  ;;    "data: [DONE]\n\n". Thus, `callback' is called with the value of `copilot-chat-magic', and
+  ;;    the two trailing empty lines are skipped.
   (when copilot-chat-last-data
     (setq string (concat copilot-chat-last-data string))
     (setq copilot-chat-last-data nil))
 
   (let ((segments (split-string string "\n")))
     (dolist (segment segments)
-      (when (string-prefix-p "data:" segment)
-        (let ((data (substring segment 6)))
-          (if (string= data "[DONE]")
-			  (progn
-				(funcall callback copilot-chat-magic)
-				(setf (copilot-chat-history copilot-chat-instance) (cons (list copilot-chat-curl-answer "assistant") (copilot-chat-history copilot-chat-instance)))
-				(setq copilot-chat-curl-answer nil))
-            (condition-case err
-                (let* ((json-obj (json-parse-string data :object-type 'alist))
-                       (choices (and json-obj (alist-get 'choices json-obj)))
-                       (delta (and (> (length choices) 0) (alist-get 'delta (aref choices 0))))
-                       (token (and delta (alist-get 'content delta))))
-                  (when (and token (not (eq token :null)))
-                    (funcall callback token)
-					(setq copilot-chat-curl-answer (concat copilot-chat-curl-answer token))))
-              (json-parse-error
-				(setq copilot-chat-last-data segment))
-              (json-end-of-file
-               (setq copilot-chat-last-data segment))
-              (error
-               (message (format "erreur : %s" segment))))))))))
+      (let ((extracted (copilot-chat--extract-segment segment)))
+        (cond
+         ;; No data at all, just skip:
+         ((eq extracted 'empty)
+          nil)
+         ;; Data looks truncated, save it for the next segment:
+         ((eq extracted 'partial)
+          (setq copilot-chat-last-data segment))
+         ;; Final segment, all done:
+         ((eq extracted 'done)
+          (funcall callback copilot-chat-magic)
+	  (setf (copilot-chat-history copilot-chat-instance) (cons (list copilot-chat-curl-answer "assistant") (copilot-chat-history copilot-chat-instance)))
+	  (setq copilot-chat-curl-answer nil))
+         ;; Otherwise, JSON parsed successfully, extract .choices[0].delta.content and pass that along:
+         (extracted
+          (let* ((choices (alist-get 'choices extracted))
+                 (delta (and (> (length choices) 0) (alist-get 'delta (aref choices 0))))
+                 (token (and delta (alist-get 'content delta))))
+            (when (and token (not (eq token :null)))
+              (funcall callback token)
+	      (setq copilot-chat-curl-answer (concat copilot-chat-curl-answer token))))))))))
```

Comment:
This has just been reindented, no change. The "Hide whitespace" diff option might make this more obvious 😄